engine/render/src/render/font.cpp

```diff
@@ -317,13 +312,60 @@ namespace dmRender
     //     }
     // }
 
+
+    // From Box2D
+    inline bool IsPowerOfTwo(uint32_t x)
+    {
+        return x > 0 && (x & (x - 1)) == 0;
+    }
+
+    // From Box2D
+    inline uint32_t NextPowerOfTwo(uint32_t x)
+    {
+        x |= (x >> 1);
+        x |= (x >> 2);
+        x |= (x >> 4);
+        x |= (x >> 8);
+        x |= (x >> 16);
+        return x + 1;
+    }
+
+    static void RecreateCache(HFontMap font_map, dmGraphics::HContext graphics_context, dmRender::FontMetrics* metrics)
+    {
+        font_map->m_CacheWidth = dmMath::Max(font_map->m_CacheWidth, metrics->m_MaxWidth);
+        font_map->m_CacheHeight = dmMath::Max(font_map->m_CacheHeight, metrics->m_MaxHeight);
+
+        if (!IsPowerOfTwo(font_map->m_CacheWidth))
+            font_map->m_CacheWidth = NextPowerOfTwo(font_map->m_CacheWidth);
+        if (!IsPowerOfTwo(font_map->m_CacheHeight))
+            font_map->m_CacheHeight = NextPowerOfTwo(font_map->m_CacheHeight);
+
+        font_map->m_MaxAscent = metrics->m_MaxAscent;
+        font_map->m_MaxDescent = metrics->m_MaxDescent;
+        RecreateTexture(font_map, graphics_context, font_map->m_CacheWidth, font_map->m_CacheHeight);
+        SetFontMapCacheSize(font_map, metrics->m_MaxWidth, metrics->m_MaxHeight, metrics->m_MaxAscent);
+    }
+
     static void UpdateGlyphTexture(HFontMap font_map, dmRender::FontGlyph* g, int32_t x, int32_t y, int offset_y)
     {
-        uint32_t glyph_data_compression; // E.g. FONT_GLYPH_COMPRESSION_NONE;
+        if (font_map->m_GetFontMetrics)
+        {
+            dmRender::FontMetrics font_metrics = {0};
+            uint32_t num_glyphs = font_map->m_GetFontMetrics(font_map->m_UserData, &font_metrics);
+
+            // If the texture is actually too small
+            if (num_glyphs > 0 && (font_map->m_CacheWidth < font_metrics.m_MaxWidth || font_map->m_CacheHeight < font_metrics.m_MaxHeight))
```

Comment:
The offline calculated cache size wasn't large enough to hold the glyphs of this font. (i..e 32 pixels high versus a 36 pixel high glyph)
So we need to recreate the texture.
We don't know the glyphs before adding them to the font.

Reply:
I guess a first step towards adding a packing step, or perhaps several pages.